substrate/frame/nis/src/migration.rs

```diff
@@ -0,0 +1,151 @@
+			if StorageVersion::get::<Pallet<T>>() == 0 {
+				Receipts::<T>::translate(|index, old_receipt: v0::OldReceiptRecordOf<T>| {
+					weight.saturating_accrue(T::DbWeight::get().reads_writes(1, 1));
+					log::info!(target: TARGET, "migrating reciept record expiry #{:?}", &index);
+					let new_expiry =
+						BlockConversion::convert_block_number_to_relay_height(old_receipt.expiry);
+					Some(ReceiptRecord {
+						proportion: old_receipt.proportion,
+						owner: old_receipt.owner,
+						expiry: new_expiry,
+					})
+				});
+
+				// Read old value
+				let old_summary = v0::Summary::<T>::get();
+
+				let new_last_period =
+					BlockConversion::convert_block_number_to_relay_height(old_summary.last_period);
+				// Convert to new format
+				let new_summary = SummaryRecord {
+					proportion_owed: old_summary.proportion_owed,
+					index: old_summary.index,
+					thawed: old_summary.thawed,
+					last_period: new_last_period,
+					receipts_on_hold: old_summary.receipts_on_hold,
+				};
+				log::info!(target: TARGET, "migrating summary record, current thaw period's beginning.");
+				// Write new value
+				Summary::<T>::put(new_summary);
+				// Update storage version
+				StorageVersion::new(1).put::<Pallet<T>>();
+				// Return weight (adjust based on operations)
+				weight.saturating_add(T::DbWeight::get().reads_writes(1, 3))
+			} else {
+				log::info!(target: TARGET, "nill");
+				weight
+			}
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(state: Vec<u8>) -> Result<(), TryRuntimeError> {
+			// Decode pre-upgrade state
+			let (old_receipts, old_summary): (
+				Vec<(ReceiptIndex, v0::OldReceiptRecordOf<T>)>,
+				v0::OldSummaryRecordOf<T>,
+			) = Decode::decode(&mut &state[..]).expect("pre_upgrade data must decode");
+
+			// Verify Receipts migration
+			for (index, old_receipt) in old_receipts {
+				let new_receipt =
+					Receipts::<T>::get(index).ok_or("Receipt missing after migration")?;
+
+				// Verify expiry conversion
+				let expected_expiry =
+					BlockConversion::convert_block_number_to_relay_height(old_receipt.expiry);
+				ensure!(new_receipt.expiry == expected_expiry, "Receipt expiry conversion failed");
+
+				// Verify other fields unchanged
+				ensure!(
+					new_receipt.proportion == old_receipt.proportion &&
+						new_receipt.owner == old_receipt.owner,
+					"Receipt fields corrupted"
+				);
+			}
+
+			// Verify Summary migration
+			let new_summary = Summary::<T>::get();
+			let old_summary_last_period =
+				BlockConversion::convert_block_number_to_relay_height(old_summary.last_period);
+			ensure!(
+				new_summary.last_period == old_summary_last_period,
+				"Summary conversion failed"
+			);
+
+			// Verify storage version
+			ensure!(StorageVersion::get::<Pallet<T>>() == 1, "Storage version not updated");
+
+			Ok(())
+		}
+	}
+}
+
+/// Migrate the pallet storage from `0` to `1`.
+pub type MigrateV0ToV1<T, BlockConversion> = frame_support::migrations::VersionedMigration<
```

Comment:
so everyone who uses this pallet should migrate to RC block numbers? why? this PR switches storage version to `1`, from `0`. this means this migration should be applied by all setups? what if my setup stays on Relay Chain? what if my setup stays on Parachain with system block provider?

Reply:
Working on a fix

Reply:
please review @muharem